README.md
# Curso GRÁTIS de GNU/Linux Ubuntu Server 18.04.x LTS DevOps

## 💰 Ajude o projeto Bora para Prática a continuar fazendo vídeos gratuitos para o Canal, acessando o link de doação do PagSeguro: https://pag.ae/bjlSJcH

Robson Vaamonde<br>
Procedimentos em TI: http://procedimentosemti.com.br<br>
Bora para Prática: http://boraparapratica.com.br<br>
Robson Vaamonde: http://vaamonde.com.br<br>
Facebook Procedimentos em TI: https://www.facebook.com/ProcedimentosEmTi<br>
Facebook Bora para Prática: https://www.facebook.com/boraparapratica<br>
Instagram Procedimentos em TI: https://www.instagram.com/procedimentoem<br>
YouTUBE Bora Para Prática: https://www.youtube.com/boraparapratica<br>
Linkedin Robson Vaamonde: https://www.linkedin.com/in/robson-vaamonde-0b029028/<br>

Esse curso tem como objetivo demonstrar a instalação e configuração da nova versão de Servidor da Canonical o GNU/Linux Ubuntu Server 18.04.x, essa versão foi projetada para uma fácil implementação, administração e configuração dos principais serviços de rede de utilizados no nosso dia a dia.

Projetado para suportar implementações On-Premise (Servidores Locais) ou Cloud (Nuvens), essa versão possui várias ferramentas integradas e soluções desenvolvidas pela Canonical que facilita a implementação em larga escala e suporte a Hosts Virtual e Cloud melhorados.

Para esse curso as metodologias de DevOps, utilizando o Git e Github serão empregadas de forma dinâmica, desenvolvendo scripts e automatizando todo os processos de instalação e configuração.

Ideal para Estudantes e Profissionais de Redes que desejam se aprofundar nos pré-requisitos técnicos e práticos para Implementar Servidores GNU/Linux Ubuntu Server em cenários On-Premise ou Cloud.

## **Links Oficiais do Ubuntu, Mint, pfSense, Microsoft e VirtualBOX:**
Canonical: https://canonical.com/<br>
Linux Ubuntu: https://ubuntu.com/<br>
Linux Mint: https://www.linuxmint.com/<br>
pfSense: https://www.pfsense.org/<br>
Microsoft Windows:  https://www.microsoft.com/pt-br/windows<br>
Oracle VirtualBOX: https://www.virtualbox.org/

## **Playlist do YouTUBE com todos os Vídeos do GNU/Linux Ubuntu Server 18.04.x LTS DevOps:**
Link da Playlist: https://www.youtube.com/playlist?list=PLozhsZB1lLUOjGzjGO4snI34V0zINevDm

## **APRESENTAÇÃO do CURSO de GNU/Linux Ubuntu Server 18.04.x LTS DevOps**

DevOps Github APRESENTAÇÃO do CURSO de GNU/Linux Ubuntu Server 18.04.x LTS, vídeo mostrando os detalhes do curso e informações extras.

[![Apresentação do Curso](http://img.youtube.com/vi/qvUVWUz1MmY/0.jpg)](http://www.youtube.com/watch?v=qvUVWUz1MmY "Apresentação do Curso")

Link da vídeo aula: https://www.youtube.com/watch?v=qvUVWUz1MmY

## **APRESENTAÇÃO do CENÁRIO do Curso de GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando o cenário das máquinas virtuais utilizadas no Curso de DevOps Github GNU/Linux Ubuntu Server 18.04.x LTS

[![Apresentação do Cenário](http://img.youtube.com/vi/8AaHNXcrxag/0.jpg)](http://www.youtube.com/watch?v=8AaHNXcrxag "Apresentação do Cenário")

Link da vídeo aula: https://www.youtube.com/watch?v=8AaHNXcrxag

## **COMO instalar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o GNU/Linux Ubuntu Server 18.04.x LTS no VirtualBOX. 

Versões LTS (Long-term support) do Ubuntu são recomendadas devido ao suporte de 5 anos da data de lançamento. A nova versão do GNU/Linux Ubuntu Server 18.04.1 LTS trás um novo instalador, vários recursos novos e várias tecnologias que facilitam a implementação de serviços em rede, Cloud e Segurança de forma rápida, fácil e eficiente para pequenas, médias e grandes empresas.

[![Instalação do Ubuntu Serber](http://img.youtube.com/vi/zDdCrqNhIXI/0.jpg)](http://www.youtube.com/watch?v=zDdCrqNhIXI "Instalação do Ubuntu Serber")

Link da vídeo aula: https://www.youtube.com/watch?v=zDdCrqNhIXI&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Apt-Get.sh

## **COMO instalar a versão Mini (Netboot - Netinstall) do GNU/Linux Ubuntu Server 18.04 LTS**

Vídeo mostrando os procedimentos básicos para instalar a versão Mini (Netboot - Netinstall) do GNU/Linux Ubuntu Server 18.04 LTS. A versão Mini também conhecida como Netboot ou Netinstall possui os recursos mínimos para a instalação do Server ou Desktop Ubuntu, essa versão e indicada para instalação via Rede, utilizando os Protocolos DHCP, TFTP e PXE, indicada para customizações de sistema ou servidores de virtualização e cloud.

[![Instalação da Versão Netboot](http://img.youtube.com/vi/l-AVJE3DZfw/0.jpg)](http://www.youtube.com/watch?v=l-AVJE3DZfw "Instalação da Versão Netboot")

Link da vídeo aula: https://www.youtube.com/watch?v=l-AVJE3DZfw&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Netboot.sh

## **COMO atualizar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para manter o Servidor GNU/Linux Ubuntu Server 18.04.x LTS sempre atualizado.

APT (Advanced Packaging Tool, em português Ferramenta de Empacotamento Avançada) é um conjunto de ferramentas usadas pelo GNU/Linux Debian e suas respectivas derivações, entre eles o Ubuntu, para administrar os pacotes .deb de uma forma automática, de modo que quando um programa é instalado o APT instala e/ou atualiza também todos os pacotes que são necessários para um funcionamento do programa.

[![Atualização do Ubuntu](http://img.youtube.com/vi/esnu8TAepHU/0.jpg)](http://www.youtube.com/watch?v=esnu8TAepHU "Atualização do Ubuntu")

Link da vídeo aula: https://www.youtube.com/watch?v=esnu8TAepHU&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Apt-Get.sh

## **COMO desligar e reiniciar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para desligar e reiniciar o Servidor GNU/Linux Ubuntu Server 18.04.x LTS

[![Desligando e Reiniciando o Ubuntu](http://img.youtube.com/vi/XLFhmnKr9O0/0.jpg)](http://www.youtube.com/watch?v=XLFhmnKr9O0 "Desligando e Reiniciando o Ubuntu")

Link da vídeo aula: https://www.youtube.com/watch?v=XLFhmnKr9O0&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Shutdown.sh

## **COMO configurar a PLACA DE REDE no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar as interfaces (placa de rede) no GNU/Linux Ubuntu Server 18.04.x LTS

Na versão 18.04.x as configurações de Placa de Rede não fica mais no arquivo /etc/network/interfaces, hoje o projeto utiliza o software Netplan para fazer as configurações de rede.

Netplan é um utilitário para configurar facilmente a rede em um sistema Linux. Você simplesmente cria uma descrição YAML das interfaces de rede necessárias e o que cada uma deve ser configurada para fazer. A partir desta descrição, o Netplan irá gerar todas as configurações necessárias para a ferramenta de renderização escolhida.

[![Configurando a Placa de Rede](http://img.youtube.com/vi/zSUd4k108Zk/0.jpg)](http://www.youtube.com/watch?v=zSUd4k108Zk "Configurando a Placa de Rede")

Link da vídeo aula: https://www.youtube.com/watch?v=zSUd4k108Zk

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Netplan.sh

## **COMO configurar o Hostname, Hosts e TCP Wrappers no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar os parâmetros de rede no GNU/Linux Ubuntu Server 18.04.x LTS

Configuração do Hostname (Nome do Computador), Hosts (Banco de Dados de Computadores IPv4/IPv6 e Hostname), TCP Wrappers arquivos hosts.allow (serviço e computadores/redes liberados) e hosts.deny (serviço e computadores/redes bloqueadas) do Servidor GNU/Linux Ubuntu Server 18.04.x LTS.

[![Configurando o Hostname](http://img.youtube.com/vi/J7eyb5ynjZA/0.jpg)](http://www.youtube.com/watch?v=J7eyb5ynjZA "Configurando o Hostname")

Link da vídeo aula: https://www.youtube.com/watch?v=J7eyb5ynjZA

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/HostnameHosts.sh

## **COMO configurar o LOCALE, TIMEZONE, UTC, UTF e KEYBOARD no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar o Locale, UTC, UTF e Keyboard no GNU/Linux Ubuntu Server 18.04.x LTS

Configuração do LOCALE (Localidade), Timezone (Fuso Horário), Data e Hora, sincronização da Data e Hora do Sistema Operacional e o Hardware (BIOS), sincronização da Data e Hora de Servidores NTP, configuração de Teclado ABNT-2, Configuração do UTF-8 no Console do Servidor GNU/Linux Ubuntu Server 18.04.x LTS

[![Configurando o Locale](http://img.youtube.com/vi/laiuH1zgg3U/0.jpg)](http://www.youtube.com/watch?v=laiuH1zgg3U "Configurando o Locale")

Link da vídeo aula: https://www.youtube.com/watch?v=laiuH1zgg3U&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/LocaleTimezoneKeyboard.sh 

## **COMO configurar o RAID-1 no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para adicionar, particionar, formatar, montar e configurar o RAID (Redundant Array of Inexpensive Drives) 1 (Espelhamento) no GNU/Linux Ubuntu Server 18.04.x LTS

RAID é um meio de se criar um subsistema de armazenamento composto por vários discos individuais, com a finalidade de ganhar segurança por meio da redundância de dados e desempenho. Popularmente, RAID seriam dois ou mais discos trabalhando simultaneamente para um mesmo fim, por exemplo, citando o exemplo de RAID 1 logo abaixo, serviria como um espelhamento simples, rápido e confiável entre dois discos, para se fazer uma cópia idêntica de um disco em outro. 

[![Configurando o RAID-1](http://img.youtube.com/vi/HqBu0wr0iyw/0.jpg)](http://www.youtube.com/watch?v=HqBu0wr0iyw "Configurando o RAID-1")

Link da vídeo aula: https://www.youtube.com/watch?v=HqBu0wr0iyw

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/HardDiskRaid.sh

## **COMO configurar o BONDING no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar a alta disponibilidade das interfaces de rede (placa de rede) em modo Bonding (união) no GNU/Linux Ubuntu Server 18.04.x LTS

Network Interface Bonding é um mecanismo usado em servidores Linux que consiste em vincular mais interfaces de rede físicas a fim de fornecer mais largura de banda do que uma única interface pode fornecer ou fornecer redundância de link em caso de falha de cabo. Este tipo de redundância de link possui vários nomes no Linux, como Bonding, Teaming ou Link Aggregation Groups (LAG).

Na versão 18.04.x as configurações de Placa de Rede não fica mais no arquivo /etc/network/interfaces, hoje o projeto utiliza o software Netplan para fazer as configurações de rede.

Link da vídeo aula: 

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Netplan.sh

## **COMO fazer SNAPSHOTS no VIRTUALBOX do Curso de GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para criar o Snapshot da Máquina Virtual do Oracle do VirtualBOX do Servidor GNU/Linux Ubuntu Server 18.04.x LTS, esse recurso é importante, pois se tiver alguma falha você consegue voltar a imagens anterior se perder todos os procedimentos executados anteriormente.

[![Snapshots Máquina Virtual](http://img.youtube.com/vi/Ypi9wntITpM/0.jpg)](http://www.youtube.com/watch?v=Ypi9wntITpM "Snapshots Máquina Virtual")


CHANGELOG
Autor: Robson Vaamonde
Site: www.procedimentosemti.com.br
Facebook: facebook.com/ProcedimentosEmTI
Facebook: facebook.com/BoraParaPratica
YouTube: youtube.com/BoraParaPratica
Data de criação: 29/01/2019
Data de atualização: 09/09/2021
Versão: 0.22
Testado e homologado para a versão do Ubuntu Server 18.04.x LTS x64
Kernel >= 4.15.x

============================================= ATUALIZAÇÕES DOS SCRIPTS =============================================
09/09/2021 - Robson Vaamonde - Criação do script: ntp.sh, atualização do script dnsdhcp.sh

18/07/2021 - Robson Vaamonde - Atualização do arquivo dhcpd.conf com as opções de: NTP Server e NetBIOS.

12/06/2021 - Robson Vaamonde - Criação do script tftppxe.sh, melhorias nos arquivos de configuração dhcpd.conf, hosts.allow
vsftpd.sh, vsftpd.conf e wordpress.sh.

09/06/2021 - Robson Vaamonde - Criação do script vsftpd.sh, melhorias no script webmin.sh e atualização da versão
do Zimbra no script zimbra.sh.

27/05/2021 - Robson Vaamonde - Atualização do script loganalyzer.sh para a nova versão 4.1.12 conforme solicitação de usuários
da Internet, feito melhoria no script (remoção de echo e acréscimo de \n)

23/05/2021 - Robson Vaamonde - Melhoria no script webmin.sh, acréscimo da instalação do Usermin, atualização do arquivos
pti.intra.hosts com suporte ao CNAME do Zimbra Server, criação do script zimbra.sh

20/05/2021 - Robson Vaamonde - Melhoria no script netdata.sh, acréscimo do monitoramento do Bind9 e ISC DHCP.

18/05/2021 - Robson Vaamonde - Criação do script dnsdhcp.sh, melhorias em vários scripts e arquivos confs.

12/05/2021 - Robson Vaamonde - Atualização do script e arquivos confs do bacula.sh.

04/05/2021 - Robson Vaamonde - Criação do script ocsinventory.sh, melhorias nos scripts rocketchat.sh

18/04/2021 - Robson Vaamonde - Criação do script lemp.sh, melhorias no script HardDiskRaid.sh e Netplan.sh.

24/03/2021 - Robson Vaamonde - melhoras nos scripts wordpress.sh, netdata.sh e atualização da versão do webmin no webmin.sh

23/03/2021 - Robson Vaamonde - Correção da falha de consulta de tabelas do PhpMyAdmin, aplicação do Path de Correção do arquivo
sql.lib.php e plugin_interface.lib.php no diretório /usr/share/phpmyadmin/libraries
23/03/2021 - Robson Vaamonde - Melhoria no arquivos de configuração do Wordpress, acréscimo da linha de configuração de múltiplos
endereços IP, hostname ou domain no Wordpress.

20/03/2021 - Robson Vaamonde - Melhoria no script lamp.sh

14/01/2021 - Robson Vaamonde - Substituição do comando service pelo systemctl nos script docker.sh, loganalyzer.sh, 
zoneminder.sh, owncloud.sh e bareos.sh

01/01/2021 - Robson Vaamonde - Melhorias e correções nos scripts: graylog.sh, ntopng.sh, zabbix.sh, fusion.sh, correções e 
melhorias nos arquivos de configuração: jvm.options, ntopng.conf, ntopng.start, zabbix_agentd.conf e zabbix_server.conf

27/12/2020 - Robson Vaamonde - Adicionado no arquivo hosts.allow a opção de liberação para acesso remoto do serviço do MySQL

10/12/2020 - Robson Vaamonde - Melhoria no Bloco de Loop de verificação das dependências dos scripts, deixando de uma forma
mais simples o entendimento

09/12/2020 - Robson Vaamonde - Atualização do script glpi.sh com a nova versão e adicionado novos pacotes de dependências
atualização do script fusion.sh com novas dependências e melhorias no script.

14/08/2020 - Robson Vaamonde - Criação dos scripts: fusion.sh e rocketchat.sh

03/08/2020 - Robson Vaamonde - Criação dos scripts: ansible.sh, glpi.sh, grafana.sh, graylog.sh, tomcat.sh, zabbix.sh.
03/08/2020 - Robson Vaamonde - Criação da Variável PTBR no script do loganalyzer.sh para adicionar o recurso de tradução
03/08/2020 - Robson Vaamonde - Criação do arquivo do Systemd do Portainer para inicializar automático o Container

24/07/2020 - Robson Vaamonde - Correção das novas dependências do Netdata, resolvido as falhas de dependência do Cloud.

22/07/2020 - Robson Vaamonde - Revisão de todos os scripts do repositório, atualização das versões dos aplicativos e 
correções de português nos scripts.

30/03/2020 - Robson Vaamonde - Atualização do arquivo de instalação do netdata.sh, acréscimo da dependência do libuv1-dev 

09/11/2019 - Robson Vaamonde - Atualização do arquivo de configuração sshd_config, acréscimo das linha MaxSessions 3, 
MaxStartups 5:60:10, AddressFamily inet, PrintLastLog yes

15/10/2019 - Robosn Vaamonde - Atualização do script webmin.sh, acréscimo da dependência unzip e correção da linha de versão do 
WEBMINVERSION=`echo webmin_*`;

28/02/2019 - Robson Vaamonde - Atualizado arquivo wp-config.php, acrescentado linhas para trabalhar com dois domínios ou dois
endereços IP diferentes no mesmo Wordpress;

25/02/2019 - Robson Vaamonde - Falha no script owncloud.sh, erro nas linhas de criação do banco de dados e na opção de 
descompactar o arquivo do owncloud.tar.bz2;

18/02/2019 - Robson Vaamonde - Informado falha no script webmin.sh, erro relacionado a linha 110, tinha uma barra de espaço
a mais no comando de apt -y install;

18/02/2019 - Robson Vaamonde - Informado falha no script netdata.sh, erro relacionado a linha 128, tinha uma barra de espaço
a mais no comando de apt -y install, acréscimo do bloco de verificação de dependências do MySQL e Apache2;

10/02/2019 - Robson Vaamonde - Atualização wordpress.sh, loganalyzer.sh, openfire.sh, zoneminder.sh, acréscimo do bloco de
verificação de dependências;

10/02/2019 - Robson Vaamonde - Atualização de todos os scripts, melhorias na parte de documentação dos scripts;

10/02/2019 - Robson Vaamonde - Atualização de todos os scripts, melhorias na parte de variável de arquivo de LOG;


glpi.sh
#!/bin/bash
# Autor: Robson Vaamonde
# Site: www.procedimentosemti.com.br
# Facebook: facebook.com/ProcedimentosEmTI
# Facebook: facebook.com/BoraParaPratica
# YouTube: youtube.com/BoraParaPratica
# Data de criação: 08/02/2019
# Data de atualização: 10/12/2020
# Versão: 0.07
# Testado e homologado para a versão do Ubuntu Server 18.04.x LTS x64
# Kernel >= 4.15.x
# Testado e homologado para a versão do GLPI-9.5.x
#
# GLPI (sigla em francês: Gestionnaire Libre de Parc Informatique, ou "Free IT Equipment Manager" em inglês) é um sistema
# gratuito de Gerenciamento de Ativos de TI, sistema de rastreamento de problemas e central de atendimento. Este software
# de código aberto é escrito em PHP e distribuído sob a Licença Pública Geral GNU.
#
# O GLPI é um aplicativo baseado na Web que ajuda as empresas a gerenciar seu sistema de informações. A solução é capaz de 
# criar um inventário de todos os ativos da organização e gerenciar tarefas administrativas e financeiras. As funcionalidades
# dos sistemas auxiliam os administradores de TI a criar um banco de dados de recursos técnicos, além de um gerenciamento e 
# histórico de ações de manutenções. Os usuários podem declarar incidentes ou solicitações (com base no ativo ou não) graças
# ao recurso de Helpdesk.
#
# Informações que serão solicitadas na configuração via Web do GLPI
# GLPI Setup
# Select your language: Português do Brasil: OK;
# Licença: Eu li e ACEITO os termos de licença acima: Continuar;
# Início da instalação: Instalar;
# Etapa 0: Verificação do ambiente: Continuar;
# Etapa 1: Instalação da conexão com o bando de dados:
#	SQL server(MariaDB ou MySQL): localhost
#	Usuário SQL: glpi
#	Senha SQL: glpi: Continuar;
# Etapa 2: Conexão com banco de dados: glpi: Continuar;
# Etapa 3: Iniciando banco de dados: Continuar;
# Etapa 4: Coletar dados: Continuar;
# Etapa 5: Uma última coisa antes de começar: Continuar;
# Etapa 6: A instalação foi concluída: Usar GLPI
#
# Usuário/Senha: glpi/glpi - conta do usuário administrador
# Usuário/Senha: tech/tech - conta do usuário técnico
# Usuário/Senha: normal/normal - conta do usuário normal
# Usuário/Senha: post-only/postonly - conta do usuário postonly
#
# Site oficial: https://glpi-project.org/pt-br/
#
# Vídeo de instalação do GNU/Linux Ubuntu Server 18.04.x LTS: https://www.youtube.com/watch?v=zDdCrqNhIXI
# Vídeo de instalação do LAMP Server no Ubuntu Server 18.04.x LTS: https://www.youtube.com/watch?v=6EFUu-I3
#
# Variável da Data Inicial para calcular o tempo de execução do script (VARIÁVEL MELHORADA)
# opção do comando date: +%T (Time)
HORAINICIAL=$(date +%T)
#
# Variáveis para validar o ambiente, verificando se o usuário é "root", versão do ubuntu e kernel
# opções do comando id: -u (user)
# opções do comando: lsb_release: -r (release), -s (short), 
# opões do comando uname: -r (kernel release)
# opções do comando cut: -d (delimiter), -f (fields)
# opção do shell script: piper | = Conecta a saída padrão com a entrada padrão de outro comando
# opção do shell script: acento crase ` ` = Executa comandos numa subshell, retornando o resultado
# opção do shell script: aspas simples ' ' = Protege uma string completamente (nenhum caractere é especial)
# opção do shell script: aspas duplas " " = Protege uma string, mas reconhece $, \ e ` como especiais
# opção do shell script: $() = Executa comandos numa subshell, retornando o resultado
USUARIO=$(id -u)
UBUNTU=$(lsb_release -rs)
KERNEL=$(uname -r | cut -d'.' -f1,2)
#
# Variável do caminho do Log dos Script utilizado nesse curso (VARIÁVEL MELHORADA)
# opções do comando cut: -d (delimiter), -f (fields)
# $0 (variável de ambiente do nome do comando)
LOG="/var/log/$(echo $0 | cut -d'/' -f2)"
#
# Declarando as variáveis para criação da Base de Dados do GLPI
USER="root"
PASSWORD="pti@2018"
#
# opção do comando create: create (criação), database (base de dados), base (banco de dados)
# opção do comando create: create (criação), user (usuário), identified by (identificado por - senha do usuário), password (senha)
# opção do comando grant: grant (permissão), usage (uso em | uso na), *.* (todos os bancos/tabelas), to (para), user (usuário)
# identified by (identificado por - senha do usuário), password (senha)
# opões do comando GRANT: grant (permissão), all (todos privilégios), on (em ou na | banco ou tabela), *.* (todos os bancos/tabelas)
# to (para), user@'%' (usuário @ localhost), identified by (identificado por - senha do usuário), password (senha)
# opção do comando FLUSH: flush (atualizar), privileges (recarregar as permissões)
DATABASE="CREATE DATABASE glpi;"
USERDATABASE="CREATE USER 'glpi' IDENTIFIED BY 'glpi';"
GRANTDATABASE="GRANT USAGE ON *.* TO 'glpi' IDENTIFIED BY 'glpi';"
GRANTALL="GRANT ALL PRIVILEGES ON glpi.* TO 'glpi' IDENTIFIED BY 'glpi';"
FLUSH="FLUSH PRIVILEGES;"
#
# Declarando a variável de download do GLPI (Link atualizado no dia 09/12/2020)
RELEASE="https://github.com/glpi-project/glpi/releases/download/9.5.3/glpi-9.5.3.tgz"
#
# Exportando o recurso de Noninteractive do Debconf para não solicitar telas de configuração
export DEBIAN_FRONTEND="noninteractive"
#
# Verificando se o usuário é Root, Distribuição é >=18.04 e o Kernel é >=4.15 <IF MELHORADO)
# [ ] = teste de expressão, && = operador lógico AND, == comparação de string, exit 1 = A maioria dos erros comuns na execução
clear
if [ "$USUARIO" == "0" ] && [ "$UBUNTU" == "18.04" ] && [ "$KERNEL" == "4.15" ]
	then
		echo -e "O usuário é Root, continuando com o script..."
		echo -e "Distribuição é >= 18.04.x, continuando com o script..."
		echo -e "Kernel é >= 4.15, continuando com o script..."
		sleep 5
	else
		echo -e "Usuário não é Root ($USUARIO) ou Distribuição não é >=18.04.x ($UBUNTU) ou Kernel não é >=4.15 ($KERNEL)"
		echo -e "Caso você não tenha executado o script com o comando: sudo -i"
		echo -e "Execute novamente o script para verificar o ambiente."
		exit 1
fi
#
# Verificando se as dependências do GLPI estão instaladas
# opção do dpkg: -s (status), opção do echo: -e (interpretador de escapes de barra invertida), -n (permite nova linha)
# || (operador lógico OU), 2> (redirecionar de saída de erro STDERR), && = operador lógico AND, { } = agrupa comandos em blocos
# [ ] = testa uma expressão, retornando 0 ou 1, -ne = é diferente (NotEqual)
echo -n "Verificando as dependências do GLPI Help Desk, aguarde... "
	for name in mysql-server mysql-common apache2 php
	do
  		[[ $(dpkg -s $name 2> /dev/null) ]] || { 
              echo -en "\n\nO software: $name precisa ser instalado. \nUse o comando 'apt install $name'\n";
              deps=1; 
              }
	done
		[[ $deps -ne 1 ]] && echo "Dependências.: OK" || { 
            echo -en "\nInstale as dependências acima e execute novamente este script\n";
            echo -en "Recomendo utilizar o script: lamp.sh para resolver as dependências."
            exit 1; 
            }
		sleep 5
#
# Script de instalação do GLPI no GNU/Linux Ubuntu Server 18.04.x
# opção do comando echo: -e (enable interpretation of backslash escapes), \n (new line)
# opção do comando hostname: -I (all IP address)
# opção do comando date: + (format), %d (day), %m (month), %Y (year 1970), %H (hour 24), %M (minute 60)
echo -e "Início do script $0 em: `date +%d/%m/%Y-"("%H:%M")"`\n" &>> $LOG
clear
#
echo
echo -e "Instalação do GLPI no GNU/Linux Ubuntu Server 18.04.x\n"
echo -e "Após a instalação do GLPI acessar a URL: http://`hostname -I | cut -d' ' -f1`/glpi/\n"
echo -e "Aguarde, esse processo demora um pouco dependendo do seu Link de Internet...\n"
sleep 5
#
echo -e "Adicionando o Repositório Universal do Apt, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	add-apt-repository universe &>> $LOG
echo -e "Repositório adicionado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Adicionando o Repositório Multiversão do Apt, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	add-apt-repository multiverse &>> $LOG
echo -e "Repositório adicionado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Atualizando as listas do Apt, aguarde..."
	#opção do comando: &>> (redirecionar a saída padrão)
	apt update &>> $LOG
echo -e "Listas atualizadas com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Atualizando o sistema, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando apt: -y (yes)
	apt -y upgrade &>> $LOG
echo -e "Sistema atualizado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Removendo software desnecessários, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando apt: -y (yes)
	apt -y autoremove &>> $LOG
echo -e "Software removidos com sucesso!!!, continuando com o script..."
sleep 5
#
echo -e "Instalando o GLPI, aguarde...\n"
#
echo -e "Instalando as dependências do GLPI, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando apt: -y (yes), \ (faz a função de quebra de pagina no comando apt)
	apt -y install php-curl php-gd php-intl php-pear php-imagick php-imap php-memcache php-pspell php-mysql \
	php-recode php-tidy php-xmlrpc php-xsl php-mbstring php-gettext php-ldap php-cas php-apcu libapache2-mod-php \
	php-json php-iconv php-xml php-cli xmlrpc-api-utils &>> $LOG
echo -e "Dependências instaladas com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Fazendo o download do GLPI do site Oficial, aguarde..."
	# removendo versões anteriores baixadas do GLPI
	# baixando a versão do GLPI
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando rm: -v (verbose)
	# opção do comando wget: -O (output document file)
	rm -v glpi.tgz &>> $LOG
	wget $RELEASE -O glpi.tgz &>> $LOG
echo -e "Download do GLPI feito com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Instalando o GLPI, aguarde..."
	# descompactando o GLPI
	# movendo o diretório do GLPI para o diretório da raiz do Apache
	# alterando o dono do diretório do GLPI
	# alterando as permissões do diretório do GLPI
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando tar: -j (bzip2), -x (extract), -v (verbose), -f (file)
	# opção do comando mv: -v (verbose)
	# opção do comando chown: -R (recursive), -v (verbose), www-data.www-data (user and group)
	# opção do comando chmod: -R (recursive), -v (verbose), 755 (User=RWX, Group=R-X, Other=R-X)
	tar -zxvf glpi.tgz &>> $LOG
	mv -v glpi/ /var/www/html/glpi/ &>> $LOG
	chown -Rv www-data:www-data /var/www/html/glpi/ &>> $LOG
	chmod -Rv 755 /var/www/html/glpi/ &>> $LOG
echo -e "GLPI instalado com sucesso!!!, continuando com o script..."
sleep 5
echo
#

echo -e "Criando o Banco de Dados do GLPI, aguarde..."
	# criando a base de dados do GLPI
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando mysql: -u (user), -p (password), -e (execute)
	mysql -u $USER -p$PASSWORD -e "$DATABASE" mysql &>> $LOG
	mysql -u $USER -p$PASSWORD -e "$USERDATABASE" mysql &>> $LOG
	mysql -u $USER -p$PASSWORD -e "$GRANTDATABASE" mysql &>> $LOG
	mysql -u $USER -p$PASSWORD -e "$GRANTALL" mysql &>> $LOG
	mysql -u $USER -p$PASSWORD -e "$FLUSH" mysql &>> $LOG
echo -e "Banco de Dados criado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Habilitando os recursos do Apache2 para suportar o GLPI, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando cp: -v (verbose)
	# opção do comando phpenmod: (habilitar módulos do PHP)
	# opção do comando a2enconf: (habilitar arquivo de configuração de site do Apache2)
	# opção do comando systemctl: restart (reinicializar o serviço)
	cp -v conf/glpi.conf /etc/apache2/conf-available/ &>> $LOG
	cp -v conf/glpi-cron /etc/cron.d/ &>> $LOG
	phpenmod apcu &>> $LOG
	a2enconf glpi &>> $LOG
echo -e "Recursos habilitados com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Editando o arquivo de configuração do Apache2 do GLPI, pressione <Enter> para continuar"
	read
	sleep 3
	vim /etc/apache2/conf-available/glpi.conf
	systemctl restart apache2 &>> $LOG
echo -e "Arquivo editado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Editando o arquivo de agendamento do GLPI, pressione <Enter> para continuar"
	read
	sleep 3
	vim /etc/cron.d/glpi-cron
echo -e "Arquivo editado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Instalação do GLPI Help Desk feita com Sucesso!!!."
	# script para calcular o tempo gasto (SCRIPT MELHORADO, CORRIGIDO FALHA DE HORA:MINUTO:SEGUNDOS)
	# opção do comando date: +%T (Time)
	HORAFINAL=$(date +%T)
	# opção do comando date: -u (utc), -d (date), +%s (second since 1970)
	HORAINICIAL01=$(date -u -d "$HORAINICIAL" +"%s")
	HORAFINAL01=$(date -u -d "$HORAFINAL" +"%s")
	# opção do comando date: -u (utc), -d (date), 0 (string command), sec (force second), +%H (hour), %M (minute), %S (second), 
	TEMPO=$(date -u -d "0 $HORAFINAL01 sec - $HORAINICIAL01 sec" +"%H:%M:%S")
	# $0 (variável de ambiente do nome do comando)
	echo -e "Tempo gasto para execução do script $0: $TEMPO"
echo -e "Pressione <Enter> para concluir o processo."
# opção do comando date: + (format), %d (day), %m (month), %Y (year 1970), %H (hour 24), %M (minute 60)
echo -e "Fim do script $0 em: `date +%d/%m/%Y-"("%H:%M")"`\n" &>> $LOG
read
exit 1
